### DISCUSSION TITLE
Joint edges not rendering for the first time in react

### DISCUSSION THREAD

**Discussion by Anuthama-Ganesan** [clientIO/joint#1873] [Q&A]
I have created a joint UI with attributes of cells and edges coming from a JSON file. But I see a weird behavior wherein the entire graph is not loading up for the first time. Only the cells are coming up for the first time, and if I go back in history and come back to the page again, I am able to see the edges. 

During the first time load-
![image](https://user-images.githubusercontent.com/71628906/199484413-7de0da44-1bde-41dc-9df5-e9c73f8b6a8b.png)

Going back in history and coming back to the same page(not reload) -
![image](https://user-images.githubusercontent.com/71628906/199484307-0b7574eb-1375-43e1-a3f1-af2dd6beea44.png)

This is the code snippet - 
```
useEffect(() => {
    if (!canvas.current) return;
    dispatch(genealogyAction.geUnitEquipmentMapping(selectedLane, selectedUnitRow.unitID));

    //set the paper
    const graph = new dia.Graph();
    const paper = new dia.Paper
      ({
        model: graph,
        background: {
          color: '#202020',
        },
        width: '100%',
        height: '700px',
        frozen: true,
        async: true,
        sorting: dia.Paper.sorting.EXACT,
        cellViewNamespace: shapes,
        gridSize: 20,
        perpendicularLinks: true,
      });

    const scroller = new ui.PaperScroller({
      paper,
      autoResizePaper: true,
      scrollWhileDragging: true,
      baseWidth: 100,
      baseHeight: 50,
      cursor: 'grab',
      contentOptions: {
        minWidth: 100,
        minHeight: 50,
      }
    });

    canvas.current.appendChild(scroller.el);
    scroller.render().positionContent('top-left');
    paper.fitToContent();

    var elements = [];
    setTemplate(genealogyTemplate, unitEquipmentMapping);

    //position the elements
    genealogyTemplate.map((a) => {
      elements.push(
        new shapes.standard.BorderedImage({
          id: a.id,
          position: { x: a.xOffset, y: a.yOffset },
          size: { width: 60, height: 60 },
          attrs: {
            border: {
              refWidth: '100%',
              refHeight: '100%',
              stroke: a.strokeColor,
              strokeWidth: 2,
              rx: 10,
            },
            image: {
              preserveAspectRatio: 'none',
              backgroundColor: '#202020',
            },
            label: {
              text: a.description + "\n" + a.unitID,
              fontSize: 11,
              fill: a.labelColor,
              textWrap: {
                width: 80,
                height: '100%'
              }
            },
            background: {
              fill: '#202020',
            }
          },
        })
      )
    });

    graph.addCells(elements);
    setImage(unitEquipmentMapping, elements);
    //graph.resetCells(elements);

    equipmentRelations.forEach((link) => {
      new shapes.standard.Link({
        source: { id: link.from_Id },
        target: { id: link.to_Id },
        attrs: {
          line: {
            connection: true,
            stroke: 'white',
            strokeWidth: 1,
          }
        },
        connector: {
          name: 'rounded'
        },
        router: {
          name: 'orthogonal',
          args: { step: 20 }
        },
        z: -1,
        weight: 1,
        minLen: 1,
        labelPosition: 'l',
        labelOffset: 1,
        labelSize: {
          width: 50,
          height: 30
        },
        size: {
          width: 50, height: 30
        }
      }).addTo(graph);
    });

    // graph.resetCells(elements);
    paper.setInteractivity({ elementMove: false })
    paper.scaleContentToFit();
    paper.unfreeze();

    paper.on('cell:pointerclick',
      function (cellView, evt, x, y) {
        evt.stopPropagation();
        if (cellView.model.attributes.attrs.label.fill == '#FF6F67') {
          if (cellView.model.attributes.attrs.label.text.split(/\n/)[1] != undefined) {
            var element = unitEquipmentMapping.find(e => e.unitID[0].trim() == cellView.model.attributes.attrs.label.text.split(/\n/)[1]);
            console.log('element', element);
            setSelectedStation(element);
          }
        }
      }
    );

    paper.on('blank:mousewheel', function (evt) {
      scroller.startPanning(evt);
    });

    paper.on('blank:pointerdown', function (evt) {
      scroller.startPanning(evt);
    });

    paper.on('element:mouseover', function (cellView, evt) {
      onPaperCellMouserOver(cellView, evt);
    });

  }, []);
```

I tried to debug and check if I am able to see the edges added to the graph(watching graph._edges). I can see them, but they are not rendered to the paper for the first time. 
The variables are internally a json, coming from redux state. 
Is there anything i am missing from my end?

**Comment by kumilingus** [comment]
Are the link views in the DOM? Can you inspect if there is any `.joint-link` in the document?

Is the `canvas` element in the live DOM when the paper is unfrozen?

**Reply by Anuthama-Ganesan** [reply]
Yes, while inspecting the elements, I am able to see them like this - 

![image](https://user-images.githubusercontent.com/71628906/199532738-e1219fdc-de1b-4694-a5fb-8589048413d7.png)

What could I be missing here?

**Reply by kumilingus** [reply]
Is the canvas element in the live DOM when the paper is unfrozen? It's necessary to properly render cells.

**Reply by Anuthama-Ganesan** [reply]
This is the code snippet i am using to render the page -

```
return (
    <>
      <div id="functionHolder" style={{ display: "hidden" }} onClick={function (e) { return functionArray }}></div>
      <div>
        <SelectedUnitRow />
        < br />
        <div className={"app-body runtime-container"} >
          <div style={{ backgroundColor: '#202020', /*overflowX: 'auto', /*marginLeft: "25px", paddingLeft: "20px",, /*,position:'relative',display: 'inline-block', float: 'left'*/ }} ref={canvas} />
        </div>
      </div>
    </>);
}
```
The canvas is rendered in the class name - app-body runtime-container
And I am able to see it in the DOM -
![image](https://user-images.githubusercontent.com/71628906/199541447-7316a8fe-b224-4c44-b6f7-01c6de6c7915.png)

Is this what is expected?

**Reply by kumilingus** [reply]
Check if the `canvas.current` is not hidden (directly or indirectly): `canvas.current.getBoundingClientRect()` *(width > 0)*
Check if the `cavas.current` is in the DOM: https://developer.mozilla.org/en-US/docs/Web/API/Node/isConnected

Also try to avoid using `sorting=EXACT` with `async` mode: https://github.com/clientIO/joint/issues/1320

**Reply by Anuthama-Ganesan** [reply]
I tried to check canvas in DOM -

![image](https://user-images.githubusercontent.com/71628906/199657564-d590efbc-62be-4d76-8d96-4aa104a78c1f.png)

And this is the output -

![image](https://user-images.githubusercontent.com/71628906/199657966-a11a2ed2-cebc-4125-b385-3189d22ae97f.png)

I tried to comment out sorting=EXACT, but still I am not seeing the links yet. I am not using z attributes either.

**Reply by kumilingus** [reply]
If you set `async: false`, does it help?

- `EXACT` is the default. You should use `sorting: dia.Paper.sorting.APPROX`. It should prevent a different z-index issue.
- Use  `paper.scaleContentToFit({ useModelGeometry: true });` in the async mode (the views might not be rendered yet, so you need to read the sizes from the model).
- When you inspect a link in the DOM. What is the `d` attribute?

**Reply by Anuthama-Ganesan** [reply]
I set async as false, didn't help.

I used sorting as APPROX along with async true, that didn't work. 

`paper.scaleContentToFit({ useModelGeometry: true });` - I tried this, but that doesn't render the diagram at all, not sure why. 

I inspected the d attribute - 

```
<g model-id="cd64552f-c5ec-436f-9b69-e7ac53d6d38c" data-type="standard.Link" id="j_26" class="joint-cell joint-type-standard joint-type-standard-link joint-link joint-theme-default">
<path fill="none" cursor="pointer" stroke="transparent" stroke-linecap="round" joint-selector="wrapper" id="v-154" stroke-width="10" stroke-linejoin="round" d="M 0 0 L -12 0 C -18.666666666666664 0 -22 -3.333333333333333 -22 -10 L -22 -12 C -22 -18.666666666666664 -18.666666666666664 -22 -12 -22 L -10 -22 C -3.333333333333333 -22 0 -18.666666666666664 0 -12 L 0 0"></path>
<path fill="none" pointer-events="none" joint-selector="line" id="v-153" stroke="white" stroke-width="1" stroke-linejoin="round" d="M 0 0 L -12 0 C -18.666666666666664 0 -22 -3.333333333333333 -22 -10 L -22 -12 C -22 -18.666666666666664 -18.666666666666664 -22 -12 -22 L -10 -22 C -3.333333333333333 -22 0 -18.666666666666664 0 -12 L 0 0" marker-end="url(#v-2-956925535)"></path>
</g>
```

Should I change the z value in each of the link object that I create?

**Reply by kumilingus** [reply]
It's nothing to do with `z` index. The `sorting=APPROX` advice was to prevent a different error - #1320.

Please make sure you provide the shape namespace for the graph too `const graph = new dia.Graph({}, { cellNamespace: shapes });`

This is how your links look like when first rendered.
<img width="89" alt="image" src="https://user-images.githubusercontent.com/3967880/199756145-f8f3a24c-f91b-439d-aa44-caafe1b4f40d.png">
It seems that the elements are on top of each other (it's a loop link).


It is hard to say what's wrong. Could you create an example of this that we can run?

It seems that you copied the code from a different places, which some does not make sense in this context (defining `size` for links, defining `labelSize` and `labelPosition` for links and not to use `DirectedGraph` layout, calling `paper.fitToContent()` and having no cells in the graph).

**Reply by Anuthama-Ganesan** [reply]
This issue is resolved. I had changed sorting to `sorting: dia.Paper.sorting.APPROX`. It solved the issue. And there seemed to have been an issue in my redux store in picking up the initial state.
